javalib/src/main/scala/java/util/TreeSet.scala

```diff
@@ -0,0 +1,212 @@
+    addAll(sortedSet)
+  }
+
+  private implicit object BoxOrdering extends Ordering[Box[E]] {
+    def compare(a: Box[E], b:Box[E]): Int = {
+      if (_comparator ne null) _comparator.compare(a.inner, b.inner)
+      else defaultOrdering[E].compare(a.inner, b.inner)
+    }
+  }
+
+  private val inner: mutable.TreeSet[Box[E]] = new mutable.TreeSet[Box[E]]()
+
+  def iterator(): Iterator[E] = {
+    new Iterator[E] {
+      private val iter = inner.clone.iterator
+
+      private var last: Option[E] = None
+
+      def hasNext(): Boolean = iter.hasNext
+
+      def next(): E = {
+        last = Some(iter.next().inner)
+        last.get
+      }
+
+      def remove(): Unit = {
+        if (last.isEmpty) {
+          throw new IllegalStateException()
+        } else {
+          last.foreach(self.remove(_))
+          last = None
+        }
+      }
+    }
+  }
+
+  def descendingIterator(): Iterator[E] = {
+    new Iterator[E] {
+      private val iter = inner.iterator.toList.reverse.iterator
+
+      private var last: Option[E] = None
+
+      def hasNext(): Boolean = iter.hasNext
+
+      def next(): E = {
+        val nxt = iter.next().inner
+        last = Some(nxt)
+        nxt
+      }
+
+      def remove(): Unit = {
+        if (last.isEmpty) {
+          throw new IllegalStateException()
+        } else {
+          last.foreach(self.remove(_))
+          last = None
+        }
+      }
+    }
+  }
+
+  def descendingSet(): NavigableSet[E] = {
+    val descSetFun = { () =>
+      val retSet = new mutable.TreeSet[Box[E]]()(BoxOrdering.reverse)
+      retSet.addAll(inner)
+      retSet
+    }
+    new NavigableView(this, descSetFun)
+  }
+
+  def size(): Int =
+    inner.size
+
+  override def isEmpty(): Boolean =
+    inner.headOption.isEmpty
+
+  override def contains(o: Any): Boolean =
+    inner.contains(Box(o.asInstanceOf[E]))
+
+  override def add(e: E): Boolean =
+    inner.add(Box(e))
+
+  override def remove(o: Any): Boolean =
+    inner.remove(Box(o.asInstanceOf[E]))
+
+  override def clear(): Unit =
+    inner.clear()
+
+  override def addAll(c: Collection[_ <: E]): Boolean = {
+    val iter = c.iterator()
+    var changed = false
+    while (iter.hasNext)
+      changed = add(iter.next()) || changed
+    changed
+  }
+
+  override def removeAll(c: Collection[_]): Boolean = {
+    val iter = c.iterator()
+    var changed = false
+    while (iter.hasNext)
+      changed = inner.remove(Box(iter.next).asInstanceOf[Box[E]]) || changed
+    changed
+  }
+
+  def subSet(fromElement: E, fromInclusive: Boolean, toElement: E,
+      toInclusive: Boolean): NavigableSet[E] = {
+    val boxedFrom = Box(fromElement)
+    val boxedTo = Box(toElement)
+    val subSetFun = { () =>
+      val toTs = 
+        if (toInclusive) inner.to(boxedTo)
+        else inner.until(boxedTo)
+
+      if (fromInclusive) toTs.from(boxedFrom)
+      else toTs.from(boxedFrom) - boxedFrom
+    }
+
+    new NavigableView(this, subSetFun)
```

Comment:
This `NavigableSet` should throw an `IllegalArgumentException` when there is an access outside of the bounds. It might just be a matter of overriding the `add` method.